SIMSBIOHUB-424: Support Artifact Intake + Misc Enhancements

NickPhura opened the pull request:
## Links to Jira Tickets

https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-424

## Description of Changes

Added a call to clamav to app.ts, alongside the multer middleware, so that we shouldn't have to call it in the endpoint anymore. Endpoint code should just be able to assume any files getting that far have already been checked and are safe.

Update submission/intake endpoint/service/repo
- Updates to better traverse the submission json structure, inserting each submission feature.
  - Now also correctly assigns `parent_submission_feature_id` when inserting submission features

Update submission search indexing
- Previously, checked all feature properties against all incoming features.
- Now, only checks feature properties registered to a given feature type, against incoming features of the same type.
- Some room for optimizing the code here a bit, as right now it calls `.find` repeatedly, when it probably should just pre-parse all of the codes ahead of time, and then do direct object de-referencing. It was doing this before, but the codes response object got more complicated and I just skipped optimizing it for now in order to get this PR out quicker.

Update codes endpoint/service/repo
- Updated api types, and modified endpoint response
- Updated app types accordingly

Update artifact/intake endpoint/service/repo

Add `required_value` column to feature_type_property table to indicate if properties are required or optional when running validation.

Added `additionalProperties: false` to most endpoints, and fixed response validation issues.

Update get submission features endpoint to replace the original `data` object with one populated from `search_*` rows.
- Added related service/repo code

Update submission patch function to also unpublish all other submissions for the same `uuid` on `publish=true`

## Testing Notes

If running both SIMS and BioHub branches, should be able to publish a survey that has 0 or more attachments. Attachments should have their corresponding submission feature records created, and matching search_string s3_key records added, and the file should be uploaded the S3.

Related PR for the SIMS changes: https://github.com/bcgov/biohubbc/pull/1193

NickPhura committed:
* Support artifact intake.
- Update submission intake endpoint and related service/repo functions
- Update artifact intake endpoint and related service/repo functions
- Update database schema to add source_id and uuid columns

NickPhura committed:
Fix migrations, and some code smells

NickPhura committed:
Add back source_system column.
Fix submission intake

NickPhura committed:
Update submission intake: validation and submission feature insert steps

NickPhura committed:
Fix submission intake

NickPhura committed:
Fix codes types/endpoint validation errors

NickPhura committed:
Fix most tests

NickPhura committed:
Fix bugs, fix tests

NickPhura committed:
Merge branch 'dev' into SIMSBIOHUB-424

NickPhura committed:
Fix tests after merging in dev

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-424

NickPhura committed:
Run fix

NickPhura commented:
Working on unit tests:

artifact-service.ts (Nick - done)
~artifact-repository.ts~

code-repository.ts (Curtis - done)

submission-service.ts (Nick - done)
submission-repository.ts (Nick - done)

search-index-repository.ts (Curtis - done)

artifact/intake.ts  (Nick - done)

NickPhura committed:
Add artifact/intake.test.ts

NickPhura committed:
Update artifact-service.test.ts

NickPhura committed:
Add submission-service.test.ts

curtisupshall committed:
SIMSBIOHUB-424: Added tests for searchIndexRepository

curtisupshall committed:
Merge branch 'SIMSBIOHUB-424' of github.com:bcgov/biohubbc-platform into SIMSBIOHUB-424

curtisupshall committed:
Added test todo

NickPhura committed:
Add more submission-service.test.ts

NickPhura committed:
Add submission-repository.test.ts

NickPhura committed:
Merge branch 'SIMSBIOHUB-424' of https://github.com/bcgov/biohubbc-platform into SIMSBIOHUB-424

curtisupshall commented on api/src/repositories/submission-repository.ts:
```diff
@@ -1454,6 +1465,88 @@ export class SubmissionRepository extends BaseRepository {
+          .where(
+            knex.raw('LOWER(name)'),
+            'IN',
+            featureTypeNames.map((featureTypeName) => featureTypeName.toLowerCase())
```
Is this syntactically correct? I thought you had to wrap the `WHERE IN` values in brackets, e.g. `WHERE LOWER(name) IN (?,?,?)`

curtisupshall replied:
On second thought, I'm guessing this is handled by Knex 🤔 Feel free to resolve this comment

NickPhura replied:
Sorry, yeah I believe knex is cool with it. It did take a few trys to get it right though. I tried using `whereIn` which you would think would work exactly like the current implementation, just minus the `IN` param. But the type checking kept throwing errors. Changing it to a regular `Where` + the `IN` param made it happy.

curtisupshall commented on api/src/services/artifact-service.ts:
```diff
@@ -49,59 +52,47 @@ export class ArtifactService extends DBService {
-    });
+    defaultLog.debug({ label: 'uploadSubmissionFeatureArtifact', message: 'S3 key', artifactS3Key });
+
+    // TODO add api codes cache: so lookups like this are fast (especially since codes dont change often)
```
Curious to hear your thoughts on how we might want to achieve query caching 🙂

NickPhura replied:
Yeah there are a few things we can do, depending on what is most helpful and/or the development cost:

Backend caching of just static data (like codes) by just fetching them on server start, so the endpoint/service can respond immediately. Could do the same on the frontend for static data as well (which we might already be achieving with the codes context?)

Frontend caching of all/some requests (using something like react-query and/or axios-cache-interceptor).

Backend caching of all/some requests (leveraging eTags and 304 statuses) (Definitely the most intense option).

curtisupshall commented on api/src/utils/file-utils.ts:
```diff
@@ -272,15 +272,15 @@ export async function getObjectMeta(key: string): Promise<HeadObjectOutput> {
-  keyParts.push('datasets');
-  keyParts.push(options.datasetUUID);
-  keyParts.push('artifacts');
+  keyParts.push('submissions');
```
Can we simplify this function a bit:

```javascript
return [
  'datasets',
  options.datasetUUID,
  // ...
].filter(Boolean).join('/');
```

NickPhura replied:
For sure, the main reason for it being separate pushes, historically, was that we had some if statements in there that would change the key depending on which fields were provided. Something we borrowed from SIMS. Though in BioHub it looks like we never had any optional fields, so there were no conditional key parts.

al-rosenthal reviewed (commented):
2 things I noticed:
- an import in `useSubmissionDataGridColumns` is upset about a missing interface, not sure if it was intentionally removed
- s3_key needs to be added to `validateProperties` in the `validation-service`

On that topic, does it make more sense to name this key vs s3_key? I'm also wondering if the key needs it's own property type since it's just another string.

NickPhura commented:
> 2 things I noticed:
> 
>     * an import in `useSubmissionDataGridColumns` is upset about a missing interface, not sure if it was intentionally removed
> 
>     * s3_key needs to be added to `validateProperties` in the `validation-service`
> 
> 
> On that topic, does it make more sense to name this key vs s3_key? I'm also wondering if the key needs it's own property type since it's just another string.

Yeah a few issues have cropped up after Mac and I's code got combined. The primary resolution is to leverage that 'calculated' column in the properties table. The validation can use it to skip validating calculated fields (currently it treats all properties as required, which breaks when it gets to s3_keys as they aren't generated until after validation has run, and also won't be saved to the properties column at all but will be added to the search_strings table directly), and then the frontend can still use it to understand how to render 's3_key' fields which will be part of the search_* table results that the frontend uses to render the data tables.

Re: 'key' vs 's3_key' - Yeah we can change it. I think I originally just landed on s3_key for no other reason than it would be easy to change later, and all of the existing artifact code uses 's3' in its naming. But, we can certainly change it now to whatever we think is best. 'artifact_key' might even be better? Just so it describes itself a bit but is still generic?

NickPhura committed:
Add test. Fix multer clamav scan

NickPhura committed:
Remove comments

al-rosenthal committed:
fixed import

al-rosenthal committed:
fixed type checking

curtisupshall committed:
Merge branch 'dev' into SIMSBIOHUB-424

al-rosenthal committed:
fixed import issue

NickPhura committed:
Remove source_id from submission.
Make uuid a source submitted field for submission.
make source_id optional for submission_features.

NickPhura committed:
Merge branch 'SIMSBIOHUB-424' of https://github.com/bcgov/biohubbc-platform into SIMSBIOHUB-424

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-424

al-rosenthal committed:
Merge branch 'dev' into SIMSBIOHUB-424

al-rosenthal committed:
fixed seed file, moved region inserts into migration

NickPhura committed:
Fix submission intake.
Add required_value column to feature_type_property to mark feature properties as required vs optional.
Update validation logic to account for required vs optional properties.
Add env var toggle to all seed files.

NickPhura committed:
Merge branch 'SIMSBIOHUB-424' of https://github.com/bcgov/biohubbc-platform into SIMSBIOHUB-424

NickPhura committed:
Fix submission ingest not assigning parent submission feature id.
Adjust get unreviewed/completed/published submission SQL.
Misc comment additions

NickPhura committed:
ignore-skip

NickPhura committed:
Skip deprecated test

NickPhura committed:
Update getFeatures endpoint to return search keys as data, in place of original raw properties.

NickPhura committed:
All of the updates/fixes/enhancements

NickPhura committed:
Fix seeding so it no longer creates submissions with no features (which is invalid)

NickPhura committed:
Fix 2 more endpoint openapi specs.
Unpublish other submissions on publish patch operation.

NickPhura committed:
ignore-skip

NickPhura committed:
Remove unused code

curtisupshall commented on api/src/paths/administrative/security/submission/{submissionId}.ts:
```diff
@@ -51,18 +51,59 @@ GET.apiDoc = {
+                  nullable: true
+                },
+                update_user: {
+                  type: 'string',
```
Wouldn't `update_user` have the same type as `create_user`?

NickPhura replied:
Yup, will fix

NickPhura committed:
PR comments

curtisupshall reviewed (approved):
Code looks good to me!

NickPhura committed:
Merge branch 'dev' into SIMSBIOHUB-424

al-rosenthal reviewed (approved):
ran this locally and things are working as expected 🌯
